full-node/db/sov-schema-db/src/snapshot.rs

```diff
@@ -0,0 +1,158 @@
+//! Snapshot related logic
+
+use std::sync::{Arc, LockResult, Mutex, RwLock, RwLockReadGuard};
+
+use crate::schema::{KeyCodec, ValueCodec};
+use crate::{Operation, Schema, SchemaBatch, DB};
+
+/// Id of database snapshot
+pub type SnapshotId = u64;
+
+/// A trait to make nested calls to several [`Schema`]
+pub trait QueryManager {
+    /// Get a value from snapshot or its parents
+    fn get<S: Schema>(
+        &self,
+        snapshot_id: SnapshotId,
+        key: &impl KeyCodec<S>,
+    ) -> anyhow::Result<Option<Operation>>;
+}
+
+/// Simple wrapper around `RwLock` that only allows read access.
+pub struct ReadOnlyLock<T> {
+    lock: Arc<RwLock<T>>,
+}
+
+impl<T> ReadOnlyLock<T> {
+    /// Create new [`ReadOnlyLock`] from [`Arc<RwLock<T>>`].
+    pub fn new(lock: Arc<RwLock<T>>) -> Self {
+        Self { lock }
+    }
+
+    /// Acquires a read lock on the underlying `RwLock`.
+    pub fn read(&self) -> LockResult<RwLockReadGuard<'_, T>> {
+        self.lock.read()
+    }
+}
+
+/// Wrapper around [`DB`] that allows to read from snapshots
+pub struct DbSnapshot<Q> {
+    id: SnapshotId,
+    cache: Mutex<SchemaBatch>,
+    manager: ReadOnlyLock<Q>,
+    db_reader: Arc<DB>,
+}
+
+impl<Q: QueryManager> DbSnapshot<Q> {
+    /// Create new [`DbSnapshot`]
+    pub fn new(id: SnapshotId, manager: ReadOnlyLock<Q>, db_reader: Arc<DB>) -> Self {
+        Self {
+            id,
+            cache: Mutex::new(SchemaBatch::default()),
+            manager,
+            db_reader,
+        }
+    }
+
+    /// Get a value from current snapshot, its parents or underlying database
+    pub fn read<S: Schema>(&self, key: &impl KeyCodec<S>) -> anyhow::Result<Option<S::Value>> {
+        // Some(Operation) means that key was touched,
+        // but in case of deletion we early return None
+        // Only in case of not finding operation for key,
+        // we go deeper
+
+        // 1. Check in cache
+        if let Some(operation) = self
+            .cache
+            .lock()
```

Comment:
Is is expected that the lock is held until the end of the scope (method in this case)

Reply:
Alternatively we could add another method `self.read` similar to: `self.put` and `self.delete` bellow and do the locking there.